src/unitxt/metrics.py

```diff
@@ -399,31 +541,39 @@ def process(self, stream: Stream, stream_name: Optional[str] = None) -> Generato
 
             instances.append(instance)
 
-        for reduction, fields in self.reduction_map.items():
-            assert (
-                reduction in self.implemented_reductions
-            ), f"Reduction {reduction} is not implemented, use one of {self.implemented_reductions}"
+        return instances, global_score
 
-            if reduction == "mean":
-                from statistics import mean
+    @staticmethod
+    def aggregate(instances, field_name):
+        scores = [instance["score"]["instance"][field_name] for instance in instances]
+        return np.nanmean(scores)
 
-                for field_name in fields:
-                    scores = [
-                        instance["score"]["instance"][field_name]
-                        for instance in instances
-                    ]
-                    global_score[field_name] = mean(scores)
-                    if field_name == self.main_score:
-                        global_score["score"] = global_score[field_name]
-                        global_score["score_name"] = self.main_score
+    def grouped_aggregate(self, instances, field_name, aggregation_func):
+        from collections import defaultdict
 
-                confidence_interval = self.score_based_confidence_interval(
-                    instances=instances
+        group_to_instance_scores = defaultdict(list)
+        for instance in instances:
+            additional_inputs = instance["additional_inputs"]
+            if self.grouping_field not in additional_inputs:
+                raise ValueError(
+                    f"Missing '{self.grouping_field}' from instance {instance}. "
+                    f"This field is required for group based metric computation."
                 )
-                global_score.update(confidence_interval)
+            group_key = additional_inputs[
+                self.grouping_field
+            ]  # do we need to convert to str?
+            group_to_instance_scores[group_key].append(
+                instance["score"]["instance"][field_name]
+            )
 
-        for instance in instances:
-            yield instance
+        group_total_scores = [
+            aggregation_func(scores) for scores in group_to_instance_scores.values()
+        ]
+        group_total_scores = [
+            score for score in group_total_scores if not np.isnan(score)
+        ]
+        # ignore NaNs in aggregation
```

Comment:
The comment is not what the line below does.. should it be removed?

Reply:
added a warnings catch for the RuntimeWarning from nanmean